### PR TITLE
${json-encode} support for rawValue

### LINKED ISSUE #2856: RawValue should work with xml-encode en json-encode layout renderers
This should work if the #1332 is working:

```xml
<parameter name="@Request1" layout="${xml-encode:inner=${event-properties:value1}}" />
<parameter name="@Request2" layout="${json-encode:inner=${event-properties:value1}}" />
```

### PR DESCRIPTION
**Pull request by 304NotModified**
Fixes #2856 

todo 

- [ ] xml-encode

### CODE REVIEW

File: src/NLog/LayoutRenderers/Wrappers/JsonEncodeLayoutRendererWrapper.cs [hunk 1]
```diff
@@ -70,9 +72,22 @@ public JsonEncodeLayoutRendererWrapper()
         [DefaultValue(true)]
         public bool EscapeUnicode { get; set; }
 
+        /// <summary>
+        /// Render with use of the Raw value, if possible.
+        /// </summary>
+        public bool Raw { get; set; }
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/Wrappers/JsonEncodeLayoutRendererWrapper.cs]: not sure with this name

File: src/NLog/LayoutRenderers/Wrappers/JsonEncodeLayoutRendererWrapper.cs [hunk 2]
```diff
@@ -70,9 +72,22 @@ public JsonEncodeLayoutRendererWrapper()
...
+
         /// <inheritdoc/>
         protected override void RenderInnerAndTransform(LogEventInfo logEvent, StringBuilder builder, int orgLength)
         {
+            if (Raw && Inner.TryGetRawValue(logEvent, out var rawValue))
+            {
+                ConfigurationItemFactory.Default.JsonConverter.SerializeObject(rawValue, builder); //todo missing EscapeUnicode
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/Wrappers/JsonEncodeLayoutRendererWrapper.cs]: todo add warn for unsupported EscapeUnicode?

**Comment by snakefoot** [comment]
What is the difference between this and `${event-properties:prop_key:format=@}` ?

**Comment by 304NotModified** [comment]
> Remember that I have now restricted RawValue to only be primitive types, and not full objects.

yes, that is indeed an issue with the original idea after this.

> ${event-properties:prop_key:format=@}

good one. Haven't think of that

**Comment by 304NotModified** [comment]
> What is the difference between this and `${event-properties:prop_key:format=@}` ?

after some thinking:

- '@' won't work for XML
- this should also work for GDC etc (but not now because of the change of IRawValue there)
- This is needed to serialize ${exception} to JSON?

**Comment by snakefoot** [comment]
> '@' won't work for XML

Nope it is a special format-option from the message-template-world.

> this should also work for GDC

GdcLayoutRenderer also has support for the Format-parameter = '@'

> This is needed to serialize ${exception} to JSON?

ExceptionLayoutRenderer also has support for the Format-parameter = '@'

**Comment by 304NotModified** [comment]
close in favor of the current option (`@` format)

**Pull request closed without merging**